pkg/reconciler/knativeeventing/common/replicasenvvarstransform.go

```diff
@@ -74,13 +75,23 @@ func ReplicasEnvVarsTransform(client unstructuredGetter) mf.Transformer {
 							// Add the new env var into the existing env vars, if it is not available in the existing
 							// cluster.
 							mergedEnv = append(mergedEnv, env)
-						} else if !nameExistsEnvVarNames(env.Name, envVarNames) {
+						} else if !envVarNames.Has(env.Name) {
 							// Set the env var value, if it is available in the existing
 							// cluster, but it is not in the preserved list of the env vars.
 							mergedEnv[envIndex] = env
 						}
 					}
-					applyContainer.Env = mergedEnv
+
+					// Remove the env var, which is not in the preserved list and not in the applied manifests
+					cleanedMergedEnv := mergedEnv
+					for _, env := range mergedEnv {
```

Comment:
hmm, I don't think this is necessary at is.

What do you think about only iterating over the existing deployment once, copying all `envVarNames` vars into a `[]corev1.EnvVar` and then appending all env vars not  in `envVarNames` from the `applyContainer`, like

```golang
var vars []corev1.EnvVar

for _, env := range currentContainer.Env {
  if envVarNames.Has(env.Name) {
    vars = append(vars, env)
  }
}

for _, env := range applyContainer.Env {
  if envVarNames.Has(env.Name) {
    vars = append(vars, env)
  }
}

applyContainer.Env = vars
```

That should do everything the current code does, I believe.

Reply:
Something is missing: if the env belongs to the preserved list, but the current does not have this env, we SHOULD apply this env.